python/pybipp.cpp

```diff
@@ -77,6 +77,23 @@ auto processing_unit_to_string(BippProcessingUnit pu) -> std::string {
   throw InvalidParameterError();
 }
 
+void center_array(py::array_t<double> input_array) {
```

Comment:
Can this be implemented in the center_vector kernel? Possibly be just setting the type of the summation variable to double precision.

Reply:
You mean to create such a new kernel for the host as per on the gpu side?

Reply:
From a quick test on our gauss4 dataset.

**Actual DP:**
```
-I- I_filtered_images[lsq]:
 [[[140.137791 139.311796 138.453218 ... 138.516213 137.500123 136.468126]
  [141.019233 140.185898 139.31997  ... 138.036177 137.036865 136.021872]
  [141.875471 141.036599 140.164324 ... 137.542034 136.559831 135.56234 ]
  ...
```
**Actual SP:**
```
-I- I_filtered_images[lsq]:
 [[[140.13211  139.30193  138.4527   ... 138.51938  137.49445  136.46696 ]
  [141.00873  140.18248  139.30014  ... 138.0462   137.04138  136.02734 ]
  [141.86841  141.03528  140.15785  ... 137.54628  136.56119  135.55005 ]
  ...
```
**SP with no DP centering in pybipp and T mean (i.e. SP) in center_vector of imager:**
```
-I- I_filtered_images[lsq]:
 [[[139.56488  138.7137   137.84973  ... 144.82498  143.74251  142.66312 ]
  [140.4383   139.59644  138.71857  ... 144.35056  143.28964  142.224   ]
  [141.29398  140.44159  139.57005  ... 143.86255  142.82587  141.76802 ]
  ...
```
**SP with no DP centering in pybipp and double mean in center_vector of imager:**
```
-I- I_filtered_images[lsq]:
 [[[139.57207  138.72888  137.85483  ... 144.8319   143.74313  142.66066 ]
  [140.4343   139.6014   138.71115  ... 144.34447  143.29675  142.22685 ]
  [141.30511  140.43947  139.56937  ... 143.85529  142.82089  141.76895 ]
```
So the issue is rather to compute the mean using the DP XYZ (as stemming from Python). Once XYZ is casted to SP the mismatch arises regardless the precision of mean.
Therefore I would keep the current implementation.

Reply:
In this case we can merge as is, but long term it would be good to make sure that results are independent of the interface language.

Reply:
I added a TODO via https://github.com/epfl-radio-astro/bipp/pull/28/commits/fd3c819fc34210607669bfb8db891299006530c8  as a reminder.